hal/armv7m/imxrt/timer.c

```diff
@@ -14,59 +14,102 @@
  */
 
 #include "../../../timer.h"
+#include "../../armv7m.h"
 #include "config.h"
 #include <arch/cpu.h>
 #include <arch/interrupts.h>
 
 
+enum { gpt_cr = 0, gpt_pr, gpt_sr, gpt_ir, gpt_ocr1, gpt_ocr2, gpt_ocr3, gpt_icr1, gpt_icr2, gpt_cnt };
+
+
 static struct {
 	intr_handler_t handler;
-	volatile time_t jiffies;
+	volatile u32 upper;
 	spinlock_t sp;
 
+	volatile u32 *base;
 	u32 interval;
 } timer_common;
 
 
-static int _timer_irqHandler(unsigned int n, cpu_context_t *ctx, void *arg)
+static time_t hal_timerCyc2Us(time_t ticks)
 {
-	(void)n;
-	(void)arg;
-	(void)ctx;
+	return (ticks * 1000) / ((GPT_FREQ_MHZ * 1000) / (GPT_PRESCALER * GPT_OSC_PRESCALER));
+}
```

Comment:
```suggestion
	time_t us;

	if ((GPT_FREQ_MHZ % GPT_OSC_PRESCALER) == 0) {
		us = ticks;
	}
	else {
		us = (ticks * 1000) / ((GPT_FREQ_MHZ * 1000) / (GPT_PRESCALER * GPT_OSC_PRESCALER));
	}

	return us;
}
```
@agkaminski What do you think about such solution.
And of course similarly in the `hal_timerUs2Cyc()` function.
Maybe instead of `1000` use for example `1024`.

Reply:
I think I found a simpler solution - I used a power of 2 (because why not). Look at disasm now, looks quite good:

```
000031dc <hal_timerGetUs>:
    31dc:       b530            push    {r4, r5, lr}
    31de:       4c0c            ldr     r4, [pc, #48]   ; (3210 <hal_timerGetUs+0x34>)
    31e0:       b083            sub     sp, #12
    31e2:       f104 001c       add.w   r0, r4, #28
    31e6:       a901            add     r1, sp, #4
    31e8:       f000 f940       bl      346c <hal_spinlockSet>
    31ec:       6ae3            ldr     r3, [r4, #44]   ; 0x2c
    31ee:       69a4            ldr     r4, [r4, #24]
    31f0:       6a5d            ldr     r5, [r3, #36]   ; 0x24
    31f2:       689a            ldr     r2, [r3, #8]
    31f4:       0692            lsls    r2, r2, #26
    31f6:       d501            bpl.n   31fc <hal_timerGetUs+0x20>
    31f8:       6a5d            ldr     r5, [r3, #36]   ; 0x24
    31fa:       3401            adds    r4, #1
    31fc:       a901            add     r1, sp, #4
    31fe:       4805            ldr     r0, [pc, #20]   ; (3214 <hal_timerGetUs+0x38>)
    3200:       f000 f948       bl      3494 <hal_spinlockClear>
    3204:       4628            mov     r0, r5
    3206:       f3c4 0115       ubfx    r1, r4, #0, #22
    320a:       b003            add     sp, #12
    320c:       bd30            pop     {r4, r5, pc}
    320e:       bf00            nop
    3210:       20000008        andcs   r0, r0, r8
    3214:       20000024        andcs   r0, r0, r4, lsr #32
```

Reply:
I wrote about using the number 1024, but that then disappeared.
The number 1000 is probably more natural.
In my proposal if the condition is met there will be no `ubfx` instruction, that is, truncation of the 10 most significant bits.

Reply:
I think so, I think I saw 1024 in your comment somewhere. I didn't remove, that's for sure.

I opt to stick with the current solution. In my opinion one instruction is not worth ~5 times more C lines. 10 MSB are not really needed, we can handle ~600 year of uptime anyway.

Reply:
It's gets too convoluted IMHO. There is implicit assumption that 1 cycle equals 1 us at line 56 anyway:

> *(timer_common.base + gpt_ocr2) += timer_common.interval;

Why not make it explicit by adding `GPT_FREQ_MHZ == GPT_PRESCALER * GPT_OSC_PRESCALER` check and get rid of `hal_timerCyc2Us` and `hal_timerUs2Cyc` altogether?

Reply:
@ziemleszcz I missed that, imho I should just fix this one, I have a ready to use function to do so.